### PR TITLE
webhook: refactor to use controller-runtime server

### PR DESCRIPTION
**Pull request by LittleFox94**
**What this PR does / why we need it**:

Refactor the webhook to use the controller-runtime provided webhook.Server instead of a plain HTTP server. The biggest benefit this brings, is it automatically loading changed certificate files instead of forever using the ones present when the server was started.

**What type of PR is this?**

I'm not quite sure, I'll label this as feature but the webhook server not reloading changed certificates could also be seen as a bug.
/kind feature

**Special notes for your reviewer**:

This could have been implemented by using the controller-runtime CertWatcher manually, but that would've been more code and aligning machine-controller with the ecosystem-provided standards seems worth it.

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
webhook: changed certificate/key files are now reloaded automatically
```

**Documentation**:
```documentation
NONE
```

### CODE REVIEW

File: cmd/webhook/main.go
Comment (LittleFox94) [inline cmd/webhook/main.go:150]: technically breaks compatibility if we enforce TLS cert and key being in the same directory

File: cmd/webhook/main.go
Comment (LittleFox94) [inline cmd/webhook/main.go:167]: I'm not sure if this is actually still needed, together with the cancel-context above

**Comment by embik** [comment]
thank you for this!

/ok-to-test

**Comment by LittleFox94** [comment]
heya @embik :)
all tests green, if you don't have comments on the two threads I started on the code, wanna press that merge button for me? :)

**Comment by embik** [comment]
> heya @embik :) all tests green, if you don't have comments on the two threads I started on the code, wanna press that merge button for me? :)

Because it's the silent time of the year, I'd like to give colleagues a few more days to voice any second opinions (PR looks good from my perspective though) before merging it if that's okay by you. So ETA would be beginning of January.

File: cmd/webhook/main.go
Comment (embik) [reply cmd/webhook/main.go:167]: I think you can just do

```go
serverContext, cancelServerContext := context.WithCancel(context.Background())
defer cancelServerContext()
if err := srv.Start(serverContext); err != nil {
    log.Fatalw("Failed to start server", zap.Error(err))
}
```

File: cmd/webhook/main.go
Comment (embik) [inline cmd/webhook/main.go:161]: Okay, looks like I do have something to comment on: This shouldn't use `context.Background()` but a context that cancels on SIGINT/SIGTERM so the server shuts down gracefully. Then you can probably drop the `defer cancelServerContext()` as well that I commented on below (don't think we need a `context.WithCancel` then).

You can use  [signals.SetupSignalHandler](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager/signals#SetupSignalHandler) for that (like we do over in the [seed-controller-manager](https://github.com/kubermatic/kubermatic/blob/acf679a2dd5572a6a20168c8b6a493f02afb7cb1/cmd/seed-controller-manager/main.go#L109)).

**Review by embik — CHANGES_REQUESTED** [review]

File: cmd/webhook/main.go
Comment (LittleFox94) [reply cmd/webhook/main.go:161]: I was thinking about that, but when I run machine-controller locally on my machine, it already reacts perfectly fine to SIGINT - you sure this is needed?

Comment (embik) [reply cmd/webhook/main.go:161]: Sorry for the late reply. I'm not 100% _sure_ it's needed, but it's the proper way to handle signals and might cover more corner cases than the current code. I would prefer it, if only for consistency with the KKP code base.

File: cmd/webhook/main.go
Comment (LittleFox94) [reply cmd/webhook/main.go]: not done, according to https://github.com/kubermatic/machine-controller/pull/1751#discussion_r1439205717

File: cmd/webhook/main.go
Comment (LittleFox94) [reply cmd/webhook/main.go]: alright, changed code according to this :)

Sorry for my late reply as well, was still off the first week of the year ^^  
Have a good new year btw :)

Comment (embik) [reply cmd/webhook/main.go]: Happy new year! Hope you got to enjoy the days off :)

**Review by embik — APPROVED** [review]
/approve

thank you for the contribution!

**Pull request merged**
